packager/packager.cc

```diff
@@ -1948,40 +1940,64 @@ void validatePackagedFile(core::Context ctx, const ast::ExpressionPtr &tree) {
     ast::ConstShallowWalk::apply(ctx, enforcePrefix, tree);
 }
 
+} // namespace
+
 void Packager::findPackages(core::GlobalState &gs, absl::Span<ast::ParsedFile> files) {
     // Ensure files are in canonical order.
     // TODO(jez) Is this sort redundant? Should we move this sort to callers?
     fast_sort(files, [](const auto &a, const auto &b) -> bool { return a.file < b.file; });
 
+    Timer timeit(gs.tracer(), "packager.findPackages");
+
     // Find packages and determine their imports/exports.
     {
-        Timer timeit(gs.tracer(), "packager.findPackages");
         core::UnfreezeNameTable unfreeze(gs);
         core::packages::UnfreezePackages packages = gs.unfreezePackages();
         for (auto &file : files) {
             if (!file.file.data(gs).isPackage(gs)) {
                 continue;
             }
 
-            core::MutableContext ctx(gs, core::Symbols::root(), file.file);
-            auto pkg = createAndPopulatePackageInfo(ctx, file);
+            auto pkg = definePackage(gs, file);
             if (pkg == nullptr) {
                 // There was an error creating a PackageInfoImpl for this file, and getPackageInfo has already
                 // surfaced that error to the user. Nothing to do here.
                 continue;
             }
+            populateMangledName(gs, pkg->name);
+
             auto &prevPkg = gs.packageDB().getPackageInfo(pkg->mangledName());
             if (prevPkg.exists() && prevPkg.declLoc() != pkg->declLoc()) {
-                if (auto e = ctx.beginError(pkg->loc.offsets(), core::errors::Packager::RedefinitionOfPackage)) {
-                    auto pkgName = pkg->name.toString(ctx);
+                if (auto e = gs.beginError(pkg->loc, core::errors::Packager::RedefinitionOfPackage)) {
+                    auto pkgName = pkg->name.toString(gs);
                     e.setHeader("Redefinition of package `{}`", pkgName);
                     e.addErrorLine(prevPkg.declLoc(), "Package `{}` originally defined here", pkgName);
                 }
             } else {
+                populatePackagePathPrefixes(gs, file, *pkg);
                 packages.db.enterPackage(move(pkg));
             }
         }
     }
+
+    setPackageNameOnFiles(gs, files);
+
+    {
+        auto packages = gs.unfreezePackages();
```

Comment:
I think you need to unfreeze the name table here as well, as `populatePackageEdges` will potentially introduce new names.

Reply:
yeah I got this confused when I extracted it from another branch I'm working on where I stop creating MangledNames for packages that don't exist, meaning that resolving import and visible_to lines doesn't require entering new names

I'll add it back to this and make a note to delete it eventually.